builtin/logical/database/version_wrapper.go

```diff
@@ -0,0 +1,268 @@
+package database
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+
+	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/vault/helper/random"
+	"github.com/hashicorp/vault/sdk/database/dbplugin"
+	"github.com/hashicorp/vault/sdk/database/newdbplugin"
+	"github.com/hashicorp/vault/sdk/helper/pluginutil"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type databaseVersionWrapper struct {
+	v4 dbplugin.Database
+	v5 newdbplugin.Database
+}
+
+// newDatabaseWrapper figures out which version of the database the pluginName is referring to and returns a wrapper object
+// that can be used to make operations on the underlying database plugin.
+func newDatabaseWrapper(ctx context.Context, pluginName string, sys pluginutil.LookRunnerUtil, logger log.Logger) (dbw databaseVersionWrapper, err error) {
+	newDB, err := newdbplugin.PluginFactory(ctx, pluginName, sys, logger)
+	if err == nil {
+		dbw = databaseVersionWrapper{
+			v5: newDB,
+		}
+		return dbw, nil
+	}
+
+	legacyDB, err := dbplugin.PluginFactory(ctx, pluginName, sys, logger)
+	if err == nil {
+		dbw = databaseVersionWrapper{
+			v4: legacyDB,
+		}
+		return dbw, nil
+	}
+
+	return dbw, fmt.Errorf("invalid database version")
+}
+
+// Initialize the underlying database. This is analogous to a constructor on the database plugin object.
+// Errors if the wrapper does not contain an underlying database.
+func (d databaseVersionWrapper) Initialize(ctx context.Context, req newdbplugin.InitializeRequest) (newdbplugin.InitializeResponse, error) {
+	if !d.isNewDB() && !d.isLegacyDB() {
+		return newdbplugin.InitializeResponse{}, fmt.Errorf("no underlying database specified")
+	}
+
+	// v5 Database
+	if d.isNewDB() {
```

Comment:
More of a nit since this is an internal struct, but why not have these methods be called `isV4` and `isV5` since we're referencing by version on the fields already?

Reply:
Good catch - I originally was using `legacy` etc. terminology but missed changing these. Updated.

Reply:
@Il Il faut le faire @pcman312